src/fmtspec.jl

```diff
@@ -164,25 +164,42 @@ _srepr(x) = repr(x)
 _srepr(x::AbstractString) = x
 _srepr(x::AbstractChar) = string(x)
 _srepr(x::Enum) = string(x)
+@static if VERSION < v"1.2.0-DEV"
+    _srepr(x::Irrational{sym}) where {sym} = string(sym)
+end
 
 function printfmt(io::IO, fs::FormatSpec, x)
     cls = fs.cls
     ty = fs.typ
     if cls == 'i'
-        ix = Integer(x)
-        ty == 'd' || ty == 'n' ? _pfmt_i(io, fs, ix, _Dec()) :
-        ty == 'x' ? _pfmt_i(io, fs, ix, _Hex()) :
-        ty == 'X' ? _pfmt_i(io, fs, ix, _HEX()) :
-        ty == 'o' ? _pfmt_i(io, fs, ix, _Oct()) :
-        _pfmt_i(io, fs, ix, _Bin())
+        try
+            ix = Integer(x)
+            ty == 'd' || ty == 'n' ? _pfmt_i(io, fs, ix, _Dec()) :
```

Comment:
Code smell here - shouldn't have a bunch of copy-paste.
The `try` should only be around the `Integer` conversion also.
Also, if the value cannot be converted to some Integer type, it should probably go directly to `_pfmt_s`.

Reply:
I am afraid that I didn't completely get what you have in mind. We can't go directly to `_pfmt_s`, because we still need the formatting of the number part of the `Number`.

If we have a composed type like `Complex{Int64}` and we want to format it with `'d'`, this information needs to get passed on to the user-defined `fmt_Number()`.

So my idea was:

- we have `_pfmt_i(out::IO, fs::FormatSpec, x::Integer, op::Op) where {Op}` which does the normal formatting.
- we add `_pfmt_i(out::IO, fs::FormatSpec, x::Number, op::Op) where {Op}` as a fallback which finally calls `fmt_Number(x, f)` where f contains `_pfmt_i()` with all the necessary parameters, so that the user simply needs to apply `f()` to the number part, e.g. `f(real(x))`.

Due to this fallback, I think, it is ok to accept `ix` in `printfmt()` to be either Integer or a number type with Integer number parts.

Example:
`Integer(2.0)` is 2.
`Integer(2.0 + 3.0im)` fails, so `pyfmt("d", 2.0 + 3.0im)` would fail. Therefore, `fmt_Number()` calls `_pfmt_i(out, fs, x::Integer, op)` on the real and imaginary part separately. So `pyfmt("d", 2.0 + 3.0im)` will finally result in `2 + 3im`. If I would go directly to `_pfmt_s`, I would receive `2.0 + 3.0im`.
Same story with `_pfmt_f`, but even more important, as e.g. `float(2 - im)` already acts on the number parts individually and results in `2.0 - 1.0im`. So the type of `fx` in that case is always not predictable.

Reply:
For complex numbers, you could use the defaults for "d" based on the type `T` in `Complex{T}`,
no need to have to define a potentially unlimited number of "ComplexXXXX" types, just for setting defaults.
Because I added the capability of storing keyword information along with the formats, that also could be used to set how particular types of complex numbers are printed.

I'm looking for methods that don't just simply solve your problem for printing complex numbers, but also handle it in a more generic, extensible fashion.

Reply:
One thing you may not have noticed, the C-style formatted strings (i.e. `f"\%f(var)"` don't use the `pfmt*` functions, they use `cfmt`, which generates a formatter using `@printf`, so it won't use the `pfmt` functions you've been modifying.